flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -0,0 +1,172 @@
+ *     <li>Assigning them to downstream tasks for further processing.</li>
+ * </ol>
+ *
+ * <p>The splits to be read are forwarded to the downstream {@link StreamingReaderOperator}
+ * which can have parallelism greater than one.
+ */
+public class StreamingMonitorFunction extends RichSourceFunction<FlinkInputSplit> implements CheckpointedFunction {
+
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingMonitorFunction.class);
+
+  private static final long DUMMY_START_SNAPSHOT_ID = -1;
+
+  private final TableLoader tableLoader;
+  private final Schema projectedSchema;
+  private final List<Expression> filterExpressions;
+  private final ScanContext options;
+
+  private volatile boolean isRunning = true;
+  private transient Object checkpointLock;
+  private transient Table table;
+  private transient ListState<Long> snapshotIdState;
+  private transient long startSnapshotId;
+
+  public StreamingMonitorFunction(TableLoader tableLoader, Schema projectedSchema, List<Expression> filterExpressions,
+                                  ScanContext options) {
+    Preconditions.checkArgument(
+        options.snapshotId() == null && options.asOfTimestamp() == null,
+        "The streaming reader does not support using snapshot-id or as-of-timestamp to select the table snapshot.");
+    Preconditions.checkArgument(
+        options.endSnapshotId() == null, "The streaming reader does not support using end snapshot id.");
+    this.tableLoader = tableLoader;
+    this.projectedSchema = projectedSchema;
+    this.filterExpressions = filterExpressions;
+    this.options = options;
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    snapshotIdState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>(
+            "snapshot-id-state",
+            LongSerializer.INSTANCE));
+    tableLoader.open();
+    table = tableLoader.loadTable();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+      startSnapshotId = snapshotIdState.get().iterator().next();
+    } else {
+      Long optionStartSnapshot = options.startSnapshotId();
+      if (optionStartSnapshot != null) {
+        if (!SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), optionStartSnapshot)) {
+          throw new IllegalStateException("The option start-snapshot-id " + optionStartSnapshot +
+              " is not an ancestor of the current snapshot");
+        }
+
+        startSnapshotId = optionStartSnapshot;
+      } else {
+        startSnapshotId = DUMMY_START_SNAPSHOT_ID;
+      }
+    }
+  }
+
+  @Override
+  public void run(SourceContext<FlinkInputSplit> ctx) throws Exception {
+    checkpointLock = ctx.getCheckpointLock();
+    while (isRunning) {
+      synchronized (checkpointLock) {
+        monitorAndForwardSplits(ctx);
```

Comment:
I think this also need to check `isRunning`. Otherwise, this thread could check `isRunning` and then stop before getting the lock. Then `cancel` could run in another thread. Once `cancel` succeeds, I'm assuming that it is okay for Flink to call `snapshotState` to produce state with the current snapshot. But then this thread might wake up and start producing splits and update `startSnapshotId`.

Reply:
Think about this again,  it's true that we need to have a `isRunning` checking inside the `synchronized` block.